File: taghelper.py
#! python
# -*- coding: utf-8 -*-

import collections
import codecs
import urllib
import re

#initialise variables
toWrite = []
tagCloud = collections.Counter()
pageDescriptions = {}
tagDescriptions = {}

#get how many comics exist from the newest page
result = re.search("/comics/([0-9]*).jpg", urllib.urlopen("http://gunnerkrigg.com").read())
numberOfComics = int(result.group(1)) + 1

#add the heading html
with codecs.open('headingtaghelper.txt','r', "utf-8-sig") as fheading:
	toWrite.append(fheading.read())

#populate the comic page descriptions from the last entry in index.txt
with codecs.open('index.txt','r', "utf-8-sig") as findex:
	for line in findex:
		if line.startswith(";"):
			continue;
		field = line.split("\t")
		pageDescriptions[int(field[0])] = field[-1][1:-1]

#create the dropdown comic selection with the comic number & description
for x in reversed(xrange (1,numberOfComics)):
	toWrite.append("\n\t\t<option value='"+str(x)+"'>"+str(x))
	if (x in pageDescriptions):
		toWrite.append(": "+pageDescriptions[x])
	toWrite.append("</option>")

#create the text input and 'copy to clipboard' button, the prev/next buttons, the comic div and the start of the tag section
toWrite.append("</select><br /><br />" +
	"\n\t\t\t Description: <input type='text' id='tocopy'><input type='submit' class='copyit' value='Copy To Clipboard' data-clipboard-target='tocopy'>" +
	"\n\t\t</div>\n\t\t<div id='comics'>" +
	"\n\t\t\t<span class='previous'><< Previous</span> <span class='next'>Next >></span>" +
	"\n\t\t\t<div class='comic'></div>\n\t\t</div>\n\t\t<div id='tags'>\n\t\t\t<div id='current'>")

#populate the tag dictionary
with codecs.open("alltags.txt","r", "utf-8-sig") as ftags:
	for line in ftags:
		if line.startswith(";"):
			continue;
		field = line.split("\t")
		tagDescriptions[field[0]]=field[1]

#populate the tagcloud counter and output the 'current comic tags' section
with codecs.open('index.txt','r', "utf-8-sig") as findex:
	for line in findex:
		if line.startswith(";"):
			continue;
		field = line.split("\t")
		tagCloud.update(field[1:-1])
		for i in range(1,len(field)-1):
			tagItem = field[i].lower()
			toWrite.append("\n\t\t\t<span class='tag hidden "+tagItem+" "+field[0]+"'")
			if field[i] in tagDescriptions:
				toWrite.append(" title='"+tagDescriptions[field[i]][:-1]+"'")
			toWrite.append(">"+field[i]+"</span>")

#end the 'current comic tags' and start the tag list section, starting with the recent tags section
toWrite.append("\n\t\t</div>\n\t\t<div id='cloud'>\n\t\t\t<div class='recent'>Recent Tags: </div>")

#output the 30 most common tags
for (key,counter) in tagCloud.most_common(30):
	print key + str(counter)
	toWrite.append("\n\t\t\t<span class='tagcloud "+key+"'")
	if key in tagDescriptions:
		toWrite.append(" title='"+tagDescriptions[key][:-1]+"'")
	toWrite.append(">"+key+"("+str(counter)+") </span> ")
toWrite.append("<br />")

#initialise the alphabetical start character variable for the loop; output alphabetically sorted tags
startCharOld = ""
for tag in sorted(tagCloud):
	if tag:
		if ord(tag.upper()[0]) != startCharOld:
			toWrite.append("<br />"+tag.upper()[0]+": ")
		toWrite.append("\n\t\t\t<span class='tagcloud "+tag+"'")
		if tag in tagDescriptions:
			toWrite.append(" title='"+tagDescriptions[tag][:-1]+"'")
		toWrite.append(">"+tag+"("+str(tagCloud[tag])+") </span> ")
		startCharOld = ord(tag.upper()[0])

#add the ending html
with codecs.open('endingtaghelper.txt','r', "utf-8-sig") as fending:
	toWrite.append(fending.read())

#convert the toWrite array to a string and write that string to taghelper.htm
s = ''.join(toWrite)
with codecs.open("taghelper.htm","w", "utf-8") as ftagged:
	ftagged.write(s)

File: taghelper.js
$("document").ready(function(){
	
	// Initialise "things to copy" array, "comic number" and the zeroclipboard thingo
	var thingstocopy = []
	var comicnumber
	var comicComment = ""
	var recentTags = []
	var clip = new ZeroClipboard($(".copyit"))
	
	// Function that moves cursor to end of text field when clicked on
	function moveCaretToEnd(el) {
		if (typeof el.selectionStart == "number") {
			el.selectionStart = el.selectionEnd = el.value.length;
		} else if (typeof el.createTextRange != "undefined") {
			el.focus();
			var range = el.createTextRange();
			range.collapse(false);
			range.select();
		}
	}
	
	// Attach function that moves cursor to end of text input field
	$("#tocopy").on("focus",function(){
		moveCaretToEnd(this);
		
		// Work around Chrome's little problem
		$(this).onmouseup = function() {
			// Prevent further mouseup intervention
			moveCaretToEnd(this);
			$(this).onmouseup = null;
			return false;
		};
	})
	
	// Load the comic selected in the dropdown field
	$(".title").on("change",function(){
		$(".tag").addClass('hidden')
		$(".comic").addClass('hidden')
		
		var classComicNumber = '.'+$(this).val()
		comicComment = ""
		var comicCommentSplit = $(this).find('option[value*="'+$(this).val()+'"]').html().split(": ")
		comicComment = comicCommentSplit.slice(1).join(": ")
		var comicurl = "http://gunnerkrigg.com/?p="+String($(this).val())
		var comicframe = "<div class='comic "+$(this).val()+"'><iframe src='" + comicurl + "' width=800 height=800 /></div>"
		$(".comic").replaceWith(comicframe)
		
		$(classComicNumber).removeClass('hidden')
		
		$(".tagcloud").removeClass("highlighted")
		
		thingstocopy = []
		
		$(classComicNumber).each(function(){
			var tagname = $(this).text()
			thingstocopy.push(tagname)
		})
		comicnumber=String($(this).val())
		$("#tocopy").val(comicnumber+thingstocopy.sort().join("\t")+"\t#"+comicComment)
	})
	
	// "next" and "previous" buttons - load the next or previous page by triggering a change in the dropdown menu
	$(".previous").on("click",function(){
		var classComicNumber = $(this).closest("#comics").find(".comic").attr("class").split(" ")[1]
		$(".title").val(parseInt(classComicNumber)-1)
		$(".title").trigger("change")
	})
	
	$(".next").on("click",function(){
		var classComicNumber = $(this).closest("#comics").find(".comic").attr("class").split(" ")[1]
		$(".title").val(parseInt(classComicNumber)+1)
		$(".title").trigger("change")
	})
	
	// Put clicked tags into the "to copy" text field; remove them if that tag is already selected; add tags to the "recent" div
	$("#cloud").on("click",".tagcloud",function(){
		$(this).toggleClass("highlighted")
		var classname = $(this).attr("class").split(" ")[1]
		var indexof = thingstocopy.indexOf(classname)
		if (indexof == -1) {
			thingstocopy.push(classname)
		}
		else {
			thingstocopy.splice(indexof,1)
		}
		if (comicnumber) {
			$("#tocopy").val(comicnumber+thingstocopy.sort().join("\t")+"\t#"+comicComment)
		}
		else {
			$("#tocopy").val(thingstocopy.sort().join("\t")+"\t#"+comicComment)
		}
		if (recentTags.indexOf(classname) == -1) {
			recentTags.push(classname)
			var appendRecent = "Recent Tags: "
			for (i=0;i<recentTags.length;i++){
				var currentTag = recentTags.sort()[i]
				appendRecent+='<span class="tagcloud '+currentTag
				if ($("."+currentTag).hasClass("highlighted")) {
					appendRecent+=' highlighted'
				}
				appendRecent+='">'+recentTags.sort()[i]+'</span> '
			}
		}

		$(".recent").html(appendRecent)
	})
})